src/nnet3/decodable-batch-looped.cc

```diff
@@ -0,0 +1,620 @@
+        CuSubMatrix<BaseFloat> sub = ivectors_chunk.RowRange(i * num_ivectors, 
+                                                             num_ivectors);
+        sub.CopyFromMat(ivectors_base);
+      }
+      computer.AcceptInput("ivector", &ivectors_chunk);
+    }
+
+    CuMatrix<BaseFloat> output;
+    computer.Run();
+    computer.GetOutputDestructive("output", &output);
+
+    if (computer.Equal(temp)) {
+      if (batch_size > 1) {
+        const std::vector< CuMatrix<BaseFloat> > &matrices = computer.GetMatrices();
+        for (std::size_t i = 0; i < matrices.size(); i++) {
+          const CuMatrix<BaseFloat> &matrix = matrices[i];
+          KALDI_ASSERT(matrix.NumRows() % batch_size == 0);
+          int32 stream_num_rows = matrix.NumRows() / batch_size;
+          if (matrix.NumRows() > 0 && matrix.NumCols() > 0) {
+            CuSubMatrix<BaseFloat> submatrix1 = matrix.RowRange(0, stream_num_rows);
+            CuSubMatrix<BaseFloat> submatrix2 = matrix.RowRange(
+                stream_num_rows * (batch_size - 1), 
+                stream_num_rows);
+            CuSubMatrix<BaseFloat> submatrix3(matrix.RowData(0), 
+                                              stream_num_rows,
+                                              matrix.NumCols(),
+                                              matrix.Stride() * batch_size); 
+            CuSubMatrix<BaseFloat> submatrix4(matrix.RowData(batch_size - 1),
+                                              stream_num_rows,
+                                              matrix.NumCols(),
+                                              matrix.Stride() * batch_size);
+
+            if (submatrix1.ApproxEqual(submatrix2))
+              batch_first->push_back(true);
+            else if (submatrix3.ApproxEqual(submatrix4))
+              batch_first->push_back(false);
+            else
+              KALDI_ERR << "One of matrix which storing state for multiple "
+                        << "streams is neither batch first nor time first.";
+          }
+        }
+      }
+
+      computer.GetSnapshot(&snapshots_[batch_size]);
+      double elapsed = timer.Elapsed();
+      KALDI_WARN << "After try " << (num_chunks+1) 
+                 << " chunks, the NnetComputer for batch-size=" << batch_size
+                 << " becomes stable, it taken " << elapsed << "s.";
+      break;
+    }
+  }
+
+  return;
+}
+
+void NnetBatchLoopedComputer::Enqueue(NnetComputeRequest *request) {
+  KALDI_ASSERT(request != NULL);
+  bool first_chunk = request->first_chunk;
+  std::queue<QueueElement> &queue = first_chunk ? chunk1_queue_ : queue_;
+ 
+  std::unique_lock<std::mutex> lck(mtx_);
+  queue.push(std::make_pair(request, std::chrono::system_clock::now()));
+  if (queue.size() >= info_.opts.max_batch_size)
+    condition_variable_.notify_one();
+
+  return;
+}
+
+bool NnetBatchLoopedComputer::Continue() {
+  if (!is_working_) {
+    std::unique_lock<std::mutex> lck(mtx_);
```

Comment:
```suggestion
    std::lock_guard<std::mutex> lck(mtx_);
```

If you don't intend to transfer the ownership of `mtx_`, 
please use `std::lock_guard`.

Reply:
the std::condition_variable::wait_until(unique_lock<mutex>& lck, ...) NEED unique_lock as para, so use unique_lock instead of lock_guard for consistence